<a href="https://colab.research.google.com/github/yuanaf/assets/blob/main/Safety_Behind_the_Wheels_with_Add_Ons.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Dataset

## Import Libraries

In [2]:
import pandas as pd

## Load the Dataset

In [3]:
url = "https://drive.google.com/uc?export=download&id=1xoNhU-QjTp30RpTPKWxhf358Sp-eU83Y"
df = pd.read_csv(url)
df.head(20)
#df.tail(5)
#df.sample(10)

Age_band_of_driver Sex_of_driver   Educational_level  \
0               18-30          Male   Above high school   
1               31-50          Male  Junior high school   
2               18-30          Male  Junior high school   
3               18-30          Male  Junior high school   
4               18-30          Male  Junior high school   
5               31-50          Male             Unknown   
6               18-30          Male  Junior high school   
7               18-30          Male  Junior high school   
8               18-30          Male  Junior high school   
9               18-30          Male  Junior high school   
10              18-30          Male   Above high school   
11              31-50          Male   Above high school   
12              18-30          Male  Junior high school   
13              31-50          Male  Junior high school   
14              31-50          Male  Junior high school   
15              18-30        Female  Junior high school   
16              18-30          Male  Junior high school   
17              18-30          Male   Elementary school   
18              18-30          Male  Junior high school   
19              18-30          Male  Junior high school   

   Vehicle_driver_relation Driving_experience             Lanes_or_Medians  \
0                 Employee              1-2yr                      Unknown   
1                 Employee         Above 10yr            Undivided Two way   
2                 Employee              1-2yr                        other   
3                 Employee             5-10yr                        other   
4                 Employee              2-5yr                        other   
5                  Unknown            Unknown                      Unknown   
6                 Employee              2-5yr            Undivided Two way   
7                 Employee              2-5yr                        other   
8                 Employee         Above 10yr                        other   
9                 Employee              1-2yr            Undivided Two way   
10                   Owner              1-2yr                        other   
11                Employee         No Licence            Undivided Two way   
12                Employee              1-2yr  Double carriageway (median)   
13                Employee             5-10yr                        other   
14                Employee         Above 10yr            Undivided Two way   
15                Employee         Above 10yr                      One way   
16                   Owner             5-10yr            Undivided Two way   
17                Employee              2-5yr                        other   
18                Employee              2-5yr                        other   
19                Employee          Below 1yr                      One way   

   Types_of_Junction Road_surface_type       Light_conditions  \
0        No junction     Asphalt roads               Daylight   
1        No junction     Asphalt roads               Daylight   
2        No junction     Asphalt roads               Daylight   
3            Y Shape       Earth roads  Darkness - lights lit   
4            Y Shape     Asphalt roads  Darkness - lights lit   
5            Y Shape           Unknown               Daylight   
6           Crossing           Unknown               Daylight   
7            Y Shape     Asphalt roads               Daylight   
8            Y Shape       Earth roads               Daylight   
9            Y Shape     Asphalt roads               Daylight   
10       No junction     Asphalt roads               Daylight   
11       No junction       Earth roads               Daylight   
12       No junction     Asphalt roads               Daylight   
13       No junction     Asphalt roads               Daylight   
14       No junction     Asphalt roads               Daylight   
15           Y Shape     Asphalt roads  Darkness - lights lit   
16           Y Shap

## Feature Descriptions






Context

> This data set is collected from Addis Ababa Sub city police departments for Masters research work.


Content

> The data set has been prepared from manual records of road traffic accident of the year 2017-20. All the sensitive information have been excluded during data encoding and finally it has 32 features and 12316 instances of the accident. Then it is preprocessed and for identification of major causes of the accident by analyzing it using different machine learning classification algorithms algorithms.





Features Description

>Age_band_of_driver: Groups drivers by age.

> Sex_of_driver: Indicates driver's gender.

> Educational_level: Records driver's education level.

> Vehicle_driver_relation: Defines driver-vehicle relationship.

> Driving_experience: Captures years of driving experience.

> Lanes_or_Medians: Specifies if lanes or medians are present.

> Types_of_Junction: Describes type of junction.

> Road_surface_type: Identifies road surface type.

> Light_conditions: Records lighting conditions.

> Weather_conditions: Describes weather during the accident.

> Type_of_collision: Categorizes collision type.

> Vehicle_movement: Describes vehicle movement.

> Pedestrian_movement: Indicates pedestrian movement.

> Cause_of_accident: Lists cause of the accident.

> Accident_severity: Rates accident severity as 0, 1, or 2, where higher numbers indicate more severity.

# Mission 1: Who's on The Road?

Please answer all the questions below to help guide you through analyzing the road traffic accident data. These questions are designed to help you understand key trends, patterns, and insights, focusing specifically on the factors contributing to accidents and their severity.

## What column will you use to answer Mission 1's question?
Let's find out the columns list in our dataset.

In [ ]:
df.columns

Index(['Age_band_of_driver', 'Sex_of_driver', 'Educational_level',
       'Vehicle_driver_relation', 'Driving_experience', 'Lanes_or_Medians',
       'Types_of_Junction', 'Road_surface_type', 'Light_conditions',
       'Weather_conditions', 'Type_of_collision', 'Vehicle_movement',
       'Pedestrian_movement', 'Cause_of_accident', 'Accident_severity'],
      dtype='object')

In [ ]:
# value_counts() in pandas counts unique values in a column and sorts them by frequency.
df['Age_band_of_driver'].value_counts()

Age_band_of_driver
18-30       4271
31-50       4087
Over 51     1585
Unknown     1548
Under 18     825
Name: count, dtype: int64

## Who has more accidents, men or women?

In [ ]:
#Your answer here
df['Sex_of_driver'].value_counts()

Sex_of_driver
Male       11437
Female       701
Unknown      178
Name: count, dtype: int64

# Mission 2: What Happens After Dark?

## Do accidents happen more during the day or night?

In [4]:
df['Light_conditions'].value_counts()

Light_conditions
Daylight                   8798
Darkness - lights lit      3286
Darkness - no lighting      192
Darkness - lights unlit      40
Name: count, dtype: int64

## What percentage of accidents occur during daylight?

In [5]:
df['Light_conditions'].value_counts(normalize=True)

Light_conditions
Daylight                   0.714355
Darkness - lights lit      0.266807
Darkness - no lighting     0.015589
Darkness - lights unlit    0.003248
Name: proportion, dtype: float64

Approximately 71% of accidents occur during daylight.

# Mission 3: When Weather Goes Wild.

## Does weather play a big role in road traffic accidents?

In [6]:
df["Weather_conditions"].value_counts(normalize=True)

Weather_conditions
Normal               0.817067
Raining              0.108071
Other                0.024034
Unknown              0.023709
Cloudy               0.010149
Windy                0.007957
Snow                 0.004953
Raining and Windy    0.003248
Fog or mist          0.000812
Name: proportion, dtype: float64

The vast majority of accidents (81.7%) happen under normal weather conditions. This suggests that weather is not the primary factor in most accidents. Among abnormal weather conditions, rain is the most significant contributor, accounting for 10.8% of accidents.

## Which abnormal weather condition is the most significant contributor to accidents?

In [ ]:
#Your answer here
# Raining

# Mission 4: How Bad Is It? 🚑

## Which weather conditions have the highest percentage of accident severity (2)?

In [34]:
df.groupby("Weather_conditions")["Accident_severity"].value_counts(normalize=True)

Weather_conditions  Age_band_of_driver  Sex_of_driver  Educational_level   Vehicle_driver_relation  Driving_experience  Lanes_or_Medians                                  Types_of_Junction  Road_surface_type  Light_conditions         Type_of_collision               Vehicle_movement  Pedestrian_movement                                                            Cause_of_accident           Accident_severity
Cloudy              18-30               Female         Elementary school   Employee                 Below 1yr           other                                             Crossing           Asphalt roads      Darkness - lights lit    Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               No distancing               2                    0.008000
                                                       Junior high school  Employee                 2-5yr               Undivided Two way                                 Unknown            Asphalt roads      Daylight                 Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               No priority to vehicle      2                    0.008000
                                        Male           Above high school   Employee                 5-10yr              Two-way (divided with broken lines road marking)  O Shape            Gravel roads       Darkness - lights lit    Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               Changing lane to the left   2                    0.008000
                                                                           Owner                    1-2yr               Undivided Two way                                 Y Shape            Asphalt roads      Daylight                 Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               Moving Backward             2                    0.008000
                                                       Elementary school   Employee                 1-2yr               Two-way (divided with broken lines road marking)  Y Shape            Unknown            Daylight                 Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               Changing lane to the right  2                    0.008000
                                                                                                                                                                                                                                                                                                                                                                                                                             ...   
Windy               Unknown             Male           Elementary school   Employee                 5-10yr              Two-way (divided with broken lines road marking)  No junction        Asphalt roads      Darkness - no lighting   Vehicle with vehicle collision  Getting off       Not a Pedestrian                                                               Changing lane to the left   2                    0.010204
                                                       High school         Employee                 Above 10yr          Undivided Two way                                 Y Shape            Asphalt roads      Daylight                 Vehicle with vehicle collision  Going straight    Not a Pedestrian                                                               No priority to pedestrian   2                    0.010204
                                                       Junior high school  Employee                 Above 10yr          One way                                           No junction        Asphalt roads      Daylight                

The weather condition that causes the highest percentage of accident severity (2) is 'Raining and Windy,' with 95% of accidents being classified as severity (2).

# Challenge: Missing Values

## Are there any missing values in the dataset?

In [8]:
df.isnull()

Age_band_of_driver  Sex_of_driver  Educational_level  \
0                   False          False              False   
1                   False          False              False   
2                   False          False              False   
3                   False          False              False   
4                   False          False              False   
...                   ...            ...                ...   
12311               False          False              False   
12312               False          False              False   
12313               False          False              False   
12314               False          False              False   
12315               False          False              False   

       Vehicle_driver_relation  Driving_experience  Lanes_or_Medians  \
0                        False               False             False   
1                        False               False             False   
2                        False               False             False   
3                        False               False             False   
4                        False               False             False   
...                        ...                 ...               ...   
12311                    False               False             False   
12312                    False               False             False   
12313                    False               False             False   
12314                    False               False             False   
12315                    False               False             False   

       Types_of_Junction  Road_surface_type  Light_conditions  \
0                  False              False             False   
1                  False              False             False   
2                  False              False             False   
3                  False              False             False   
4                  False              False             False   
...                  ...                ...               ...   
12311              False              False             False   
12312              False              False             False   
12313              False              False             False   
12314              False              False             False   
12315              False              False             False   

       Weather_conditions  Type_of_collision  Vehicle_movement  \
0                   False              False             False   
1                   False              False             False   
2                   False              False             False   
3                   False              False             False   
4                   False              False             False   
...                   ...                ...               ...   
12311               False              False             False   
12312               False              False             False   
12313               False              False             False   
12314               False              False             False   
12315               False              False             False   

       Pedestrian_movement  Cause_of_accident  Accident_severity  
0                    False              False              False  
1                    False              False              False  
2                    False              False              False  
3                    False              False              False  
4                    False              False              False  
...                    ...                ...                ...  
12311                False              False              False  
12312                False              False              False  
12313                False              False              False  
12314                False              False              False  
12315                False              False              False  

[12316

In [9]:
df.isnull().sum()

Age_band_of_driver         0
Sex_of_driver              0
Educational_level          0
Vehicle_driver_relation    0
Driving_experience         0
Lanes_or_Medians           0
Types_of_Junction          0
Road_surface_type          0
Light_conditions           0
Weather_conditions         0
Type_of_collision          0
Vehicle_movement           0
Pedestrian_movement        0
Cause_of_accident          0
Accident_severity          0
dtype: int64

## How about 'Unknown' values in the dataset? Are they considered missing values?

Yes, 'Unknown' values in the dataset can be considered missing values as they represent a lack of specific information.

In [12]:
df['Age_band_of_driver'].value_counts()

Age_band_of_driver
18-30       4271
31-50       4087
Over 51     1585
Unknown     1548
Under 18     825
Name: count, dtype: int64

In [33]:
columns = df.columns
for i in columns:
  print(df[i].value_counts())
  print("\n")

Age_band_of_driver
18-30       4271
31-50       4087
Over 51     1585
Unknown     1548
Under 18     825
Name: count, dtype: int64


Sex_of_driver
Male       11437
Female       701
Unknown      178
Name: count, dtype: int64


Educational_level
Junior high school    7619
Elementary school     2163
High school           1110
Unknown                841
Above high school      362
Writing & reading      176
Illiterate              45
Name: count, dtype: int64


Vehicle_driver_relation
Employee    9627
Owner       1973
Unknown      593
Other        123
Name: count, dtype: int64


Driving_experience
5-10yr        3363
2-5yr         2613
Above 10yr    2262
1-2yr         1756
Below 1yr     1342
Unknown        829
No Licence     118
unknown         33
Name: count, dtype: int64


Lanes_or_Medians
Two-way (divided with broken lines road marking)    4411
Undivided Two way                                   3796
other                                               1660
Double carriageway (median)      